### PR TITLE
Port some intrinsic lowering code from LLVM/C++.

### PR DESCRIPTION
**Pull request by maleadt**
I'm hoping to make the LLVM back-end more generic, focussing on only the IR downgrade part (which could be used by other back-ends too, e.g., to target NVVM).

This PR also includes a fix, now properly attaching AIR intrinsic metadata where we didn't before. As a result, it'll break some broken tests in Metal.jl, which I'll fix after this is merged.

### CODE REVIEW

File: src/metal.jl [hunk 1]
```diff
@@ -774,3 +746,283 @@ function add_module_metadata!(@nospecialize(job::CompilerJob), mod::LLVM.Module)
...
+            @dispose builder=IRBuilder(ctx) begin
+                position!(builder, call)
+                debuglocation!(builder, call)
+
+                new_value = call!(builder, new_intr, arguments(call))
+                replace_uses!(call, new_value)
+                unsafe_delete!(bb, call)
+                changed = true
+            end
+        end
```
Comment (maleadt) [inline src/metal.jl:988]: @gbaraldi Look what you made me do.

Comment (gbaraldi) [reply src/metal.jl:988]: Sorry :(. It's very annoying that backends don't know how to lower this.

Comment (maleadt) [reply src/metal.jl:988]: No problem 🙂 But yeah it is annoying that you can't reasonably expect fairly basic things to be available across all back-ends; it should be easy enough for LLVM to provide a fallback lowering...

Comment (gbaraldi) [reply src/metal.jl:988]: X86 not having a lowering is telling though.

**Pull request merged**